Docs: Document monorepo structure (#18485)
Automattic/jetpack: docs/monorepo.md

## Jetpack Monorepo Overview

> Welcome to the Jetpack Monorepo! This document will give you some idea of the layout, and what is required for your project to fix in with our tooling.

Reviewer: ```suggestion
Welcome to the Jetpack Monorepo! This document will give you some idea of the layout, and what is required for your project to fit in with our tooling.
```

### Layout

Projects are divided into WordPress plugins, Composer packages, and Gutenberg editor extensions.

* WordPress plugins live in subdirectories of `projects/plugins/`. The directory name should probably match the WordPress plugin name, with a leading "jetpack-" removed if applicable.
* Composer packages live in subdirectories of `projects/packages/`. The directory name should probably match the package name with the leading "Automattic/jetpack-" removed.
* Editor extensions live in subdirectories of `projects/editor-extensions/`. The directory name should match the feature name (without a "jetpack/" prefix).

Tooling that's applicable to the monorepo as a whole, including tooling for generically handling projects, lives in `tools/`.

> 

Reviewer 2: Believe it's worth mentioning where WordPress itself is located.

```suggestion

WordPress, being a part of the Docker environment, gets installed into the directory `tools/docker/wordpress`, with non-monorepo plugins stored in `tools/docker/wordpress/wp-content/plugins`.

```

Documentation that's applicable to the monorepo as a whole lives in `docs/`.

All GitHub Actions configuration for the monorepo, including CI, lives in `.github`. We should strive to make things here generic rather than specifc to any one project.

> * Actual actions live in `.github/actions/`. If it doesn't have a `action.yml` file, it shouldn't be in there.

Reviewer 3: ```suggestion
* Actual actions live in `.github/actions/`. If it doesn't have an `action.yml` file, it shouldn't be in there.
```

> * Pattern matchers (not associated with an action), go in `.github/matchers/`.

Reviewer 3: ```suggestion
* Pattern matchers (not associated with an action) go in `.github/matchers/`.
```

* Other files specific to actions, including scripts used with `run:`, go in `.github/files/`.

### Compatibility

All projects should be compatible with PHP versions WordPress supports. That's currently PHP 5.6 to 8.0.

> We use eslint and phpcs to lint JavaScript and PHP code. Projects should comply with the standards enforced by these tools.

Reviewer 3: Should we be more specific about which coding standards projects should comply with?

Reviewer 3: Maybe just a link to the [Coding Standards](https://github.com/Automattic/jetpack/blob/master/docs/development-environment.md#coding-standards) section of the dev environment doc would be good?

Author: That sounds like a good idea to me.

* Projects may include `.eslintrc.js` to adjust eslint configuration as necessary, but try to keep to the spirit of it.
* As eslint does not support per-directory `.eslintignore`, any necessary ignore rules should be added to the file in the monorepo.
* As phpcs does not support per-directory configuration, any necessary file ignore clauses should be added to `.phpcs.config.xml`.

### Project structure

> We use `composer.json` to hold metadata about projects. Much of out generic tooling reads this metadata to customize handling of the project. Metadata keys used are:

Reviewer 3: ```suggestion
We use `composer.json` to hold metadata about projects. Much of our generic tooling reads this metadata to customize handling of the project. Metadata keys used are:
```

* `.name`: Generally "Automattic/jetpack-_something_". Used to report names in various places. For Composer packages, this must, of course, match the name on Packagist.

> * `.version`: If present, updated by `tools/plugin-version.sh`.

Reviewer 4: ```suggestion
* `.version`: If present, updated by `tools/plugin-version.sh`. This should not be included on Composer packages that will be served through Packagist.
```

* `.scripts.build-production`: If your project requires a build step, this must run the necessary commands.
* `.scripts.phpunit`: Packages must either set this to run PHPUnit tests, or must include a file `tests/php/ci-can-run.sh` that exits with a failure status.
* `.extra.mirror-repo`: This specifies the name of the GitHub mirror repo, i.e. the "Automattic/jetpack-_something_" in "https://github.com/Automattic/jetpack-_something_".
* `.extra.release-branch-prefix`: Our mirroring and release tooling considers any branch named like "_prefix_/branch-_version_" to be a release branch, and this specifies which _prefix_ belongs to the project.
* `.extra.version-constants`: When `tools/plugin-version.sh` is updating versions, this specifies PHP constants to replace. The value is an object matching constants to the file (relative to the plugin root) in which the constant is defined.
  * Note that constant definitions must be on a single line and use single quotes to be detected by the script. Like this:
    ```php
    define( 'CONSTANT', 'version' );
    ```

> * `.extra.wp-plugin-name`: Specify the WordPress.org plugin name, for use by scripts that deploy the plugin to WordPress.org.

Reviewer: I wonder if there should be a little indicator to show what fields are mandatory / recommended / optional?

Author: I wouldn't oppose that, if someone wants to do it.

> * `.extra.wp-plugin-name`: Specify the WordPress.org plugin name, for use by scripts that deploy the plugin to WordPress.org.

Reviewer 4: ```suggestion
* `.extra.wp-plugin-name`: Specify the WordPress.org plugin slug, for use by scripts that deploy the plugin to WordPress.org.
```
Suggesting to be more exact that we need their slug (e.g. wordpress.org/plugins/$SLUG)

> * `.extra.wp-plugin-name`: Specify the WordPress.org plugin name, for use by scripts that deploy the plugin to WordPress.org.

Reviewer 3: This change would make the item description match the other descriptions:

```suggestion
* `.extra.wp-plugin-name`: This specifies the WordPress.org plugin name, for use by scripts that deploy the plugin to WordPress.org.
```

4. If there are any built files in `.gitignore` that should be included in the mirror, use `.gitattributes` to tag them with "production-include".
5. Set `.extra.mirror-repo` in your project's `composer.json` to the name of the repo.
   * When you push the PR making this change to `composer.json`, pay attention to the Build workflow. Download the "jetpack-build" artifact and make sure it contains your project, and that there are no extra or missing files.

### Plugin release tooling

If you have set `.extra.mirror-repo`, `.extra.release-branch-prefix`, and `.extra.wp-plugin-name` in your plugin's `composer.json`, we have tooling to make releasing to WordPress.org easier.

* `tools/create-release-branch.sh` will help you create the correctly named release branch, and will automatically update version numbers and versions of monorepo packages for you. The GitHub Action will then mirror this branch to your plugin's mirror repo.
* `tools/deploy-to-svn.sh` will prepare a temporary directory with the content of the mirror repo branch that is ready to be pushed to WordPress.org SVN.

> * `tools/revert-release.sh` will prepare a temporary directory that updates the "Stable version" tag in `readme.txt` set to the previous version, in case an emergency rollback is required.

Reviewer 3: ```suggestion
* `tools/revert-release.sh` will prepare a temporary directory that updates the "Stable version" tag in `readme.txt` to the previous version, in case an emergency rollback is required.
```
